// react-angular/src/angularComponent.js
import angular from 'angular';
import React from 'react';
import AngularTemplate from './angularTemplate';

// This comes straight from Angular code, although Angular does not expose it.
// Didn't change the default '_' separator, although Angular itself only uses it with '-'.
const SNAKE_CASE_REGEXP = /[A-Z]/g;
function snake_case(name, separator = '_') {
  return name.replace(SNAKE_CASE_REGEXP, function(letter, pos) {
    return (pos ? separator : '') + letter.toLowerCase();
  });
}

function fromPairs(pairs) {
  return pairs.reduce((acc, [key, value]) => {
    acc[key] = value;
    return acc;
  }, {});
}

export default function angularComponent(directive, { wrap = false } = {}) {
  const directiveElement = snake_case(directive, '-');

  return class AngularComponent extends React.Component {
    constructor(props) {
      super(props);

      const propsPairs = Object.keys(props).map((key) => [key, props[key]]);
      const valueWrapperAttrs = fromPairs(propsPairs
        .filter(([, prop]) => angular.isString(prop))
        .map(([key, value]) => ['data-' + snake_case(key, '-'), value])
      );
      const expressionsScope = propsPairs
        .filter(([, prop]) => angular.isFunction(prop))
        .map(([key, value]) => ['data-' + snake_case(key, '-'), key, value])
      ;

      this.wrapperAttrs = {
        ...valueWrapperAttrs,
      };
      this.scopeValues = {
        ...expressionsScope,
      }
    }

    render() {
      return <AngularTemplate wrapperTag={directiveElement} wrapperAttrs={this.wrapperAttrs}/>
    }
  };
}
